src/Internal/Service/PaymentProcessingService.php

```diff
@@ -41,13 +56,39 @@ public function __construct( StateFactory $state_factory ) {
 	 *
 	 * @throws Exception
 	 * @throws StateTransitionException In case a state cannot be initialized.
+	 * @throws PaymentRequestException  When the request is malformed. This should be converted to a failure state.
 	 * @throws ContainerException       When the dependency container cannot instantiate the state.
 	 */
 	public function process_payment( int $order_id ) {
-		$context = new PaymentContext();
-		$state   = $this->state_factory->create_state( InitialState::class, $context );
-		$state   = $state->process();
+		// Start with a basis context.
+		$context = $this->create_payment_context( $order_id );
+
+		// Add details from the request.
+		$request = $this->create_payment_request();
+		$request->populate_context( $context );
```

Comment:
Nitpick: Should these steps be merged into one method in `PaymentProcessingService`? I don't think we need to be so details into the `process_payment` method to get the right PaymentContext.

Reply:
This is only done the way it is, to allow mocks in tests. Keeping the instantiation of both `PaymentContext` and `PaymentRequest` within `process_payment` would not only make the tests for `process_payment` enormous, but also force us to test everything together, instead of individually (units).